worker/runtime/cni_network.go

```diff
@@ -220,6 +235,63 @@ func (n cniNetwork) SetupMounts(handle string) ([]specs.Mount, error) {
 	}, nil
 }
 
+func (n cniNetwork) setupRestrictedNetworks() error {
+	err := createIptablesChain()
+	if err != nil {
+		return err
+	}
+
+	for _, network := range n.restrictedNetworks {
+		err = createRejectRules(network, "CNI-ADMIN")
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Create chain - ignore error if it already exists
+func createIptablesChain() error {
+	cmd := exec.Command("iptables", "-t", "filter", "-N", "CNI-ADMIN")
+	_, err := cmd.Output()
```

Comment:
Running a binary and parsing the output always feels kinda fragile to me

- https://godoc.org/github.com/coreos/go-iptables/iptables
- https://godoc.org/github.com/docker/libnetwork/iptables
- https://godoc.org/github.com/google/nftables (successor to iptables)

Reply:
Yeah, I agree - although, worth noting that at least the two iptables libraries above do essentially the same thing (but probably better to let them deal with it rather than us!)

Reply:
I agree that it's preferable to avoid using binaries. There are a few reasons we made an exception. 

There are exactly four calls being made to `iptables` here. The two `iptables` libraries mentioned above would result in four analogous calls with similar parameters, which would then in turn call the binary. We felt this did not give enough benefits to import an extra library. Given the specificity of the error handling and optimization we're doing in table manipulation these direct calls also offer more transparency.

The libraries themselves, while from recognized orgs, are not officially affiliated with the Linux `netfilter` which is what `iptables` is making use of. They would break in the same way our code would if there was ever a breaking change on a newer version of `iptables`.

`nftables` is a good suggestion. We looked at whether to use it versus `iptables` or `firewalld` and decided further investigation is [needed](https://github.com/concourse/concourse/issues/5136#issuecomment-646872293) which will be done in #5136 in the future.

Reply:
Worth noting that just because `iptables` is on the `PATH`, doesn't mean it'll behave consistently everywhere. Looking briefly at coreos/ip-tables, it looks like they parse the `iptables` version and behave slightly differently based on that.

https://github.com/coreos/go-iptables/blob/521ee6c1700638dc6444e0a9a7511ad3fd7d966a/iptables/iptables.go#L476-L478
```
// getIptablesVersion returns the first three components of the iptables version
// and the operating mode (e.g. nf_tables or legacy)
// e.g. "iptables v1.3.66" would return (1, 3, 66, legacy, nil)
```

According to this issue https://github.com/coreos/go-iptables/issues/49, it also looks like `nf_tables` can be used as a backend for the `iptables` command (I guess?), and that affects the output format in some cases.

My point being it seems like there's a fair bit of nuance around interfacing with `iptables` - we may not experience it with how we're using it, but I'm not confident that we never will - so maybe it's worth not taking on that complexity ourselves.

Reply:
@aoldershaw Digging into the code of `go-iptables` I see that it tries to detect the version of `iptables` but does not act upon this information in any significant way. The `mode` field which can be either `nftables` or `legacy` only gets acted upon for the where the `iptables` command is used in verbose mode (using `-v`). This command lists two new columns, `pkts` & `bytes`, which we don't have any use for.

Great point about the `PATH`! I'm not fully sure about the implications of this for the typical operator but it's something good to keep in mind.

I noticed that the [CNI network plugins](github.com/containernetworking/plugins), which we dependon,  use `go-iptables` under the hood as well. This somewhat counters my earlier point about breaking changes. We'd probably want to follow whatever versions of `go-iptables` that project is using, and if they decided it was worth updating against a breaking change might make sense for us to do so as well.

Given those two points I'm leaning on going with your suggestion of using `go-iptables`. Will let you know once the changes are in.